File: packages/web/src/lib/protocol.ts
/**
 * WebSocket chat protocol type definitions.
 * Mirrors packages/agent/src/argus_agent/api/protocol.py
 */

// --- Client → Server ---

export type ClientMessageType =
  | "user_message"
  | "action_response"
  | "cancel"
  | "ping";

export interface ClientMessage {
  type: ClientMessageType;
  id: string;
  data: Record<string, unknown>;
}

// --- Server → Client ---

export type ServerMessageType =
  | "connected"
  | "system_status"
  | "thinking_start"
  | "thinking_end"
  | "tool_call"
  | "tool_result"
  | "assistant_message_start"
  | "assistant_message_delta"
  | "assistant_message_end"
  | "action_request"
  | "action_executing"
  | "action_complete"
  | "alert"
  | "investigation_start"
  | "investigation_update"
  | "investigation_end"
  | "budget_update"
  | "error"
  | "pong";

export interface ServerMessage {
  type: ServerMessageType;
  timestamp: string;
  data: Record<string, unknown>;
}

export type ToolResultDisplayType =
  | "log_viewer"
  | "metrics_chart"
  | "process_table"
  | "json_tree"
  | "code_block"
  | "diff_view"
  | "table"
  | "text";

export interface ActionRequest {
  id: string;
  tool: string;
  description: string;
  command: string[] | string;
  risk_level: "READ_ONLY" | "LOW" | "MEDIUM" | "HIGH" | "CRITICAL";
  reversible: boolean;
  requires_password: boolean;
}

export interface AlertData {
  id: string;
  severity: string;
  title: string;
  summary: string;
  source: string;
  timestamp: string;
  investigation_thread_id?: string;
}

export interface BudgetStatus {
  hourly_used: number;
  hourly_limit: number;
  hourly_pct: number;
  daily_used: number;
  daily_limit: number;
  daily_pct: number;
  total_tokens: number;
  total_requests: number;
}

export interface InvestigationData {
  investigation_id: string;
  trigger?: string;
  severity?: string;
  content?: string;
  summary?: string;
  tokens_used?: number;
  type?: string;
}

File: packages/agent/src/argus_agent/api/protocol.py
"""WebSocket chat protocol message type definitions."""

from __future__ import annotations

from datetime import UTC, datetime
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

# --- Client → Server Messages ---


class ClientMessageType(StrEnum):
    USER_MESSAGE = "user_message"
    ACTION_RESPONSE = "action_response"
    CANCEL = "cancel"
    PING = "ping"


class ClientMessage(BaseModel):
    type: ClientMessageType
    id: str = ""
    data: dict[str, Any] = Field(default_factory=dict)


# --- Server → Client Messages ---


class ServerMessageType(StrEnum):
    CONNECTED = "connected"
    SYSTEM_STATUS = "system_status"
    THINKING_START = "thinking_start"
    THINKING_END = "thinking_end"
    TOOL_CALL = "tool_call"
    TOOL_RESULT = "tool_result"
    ASSISTANT_MESSAGE_START = "assistant_message_start"
    ASSISTANT_MESSAGE_DELTA = "assistant_message_delta"
    ASSISTANT_MESSAGE_END = "assistant_message_end"
    ACTION_REQUEST = "action_request"
    ACTION_EXECUTING = "action_executing"
    ACTION_COMPLETE = "action_complete"
    ALERT = "alert"
    INVESTIGATION_START = "investigation_start"
    INVESTIGATION_UPDATE = "investigation_update"
    INVESTIGATION_END = "investigation_end"
    BUDGET_UPDATE = "budget_update"
    ERROR = "error"
    PONG = "pong"


class ToolResultDisplayType(StrEnum):
    """Rich display types for tool results in the chat UI."""

    LOG_VIEWER = "log_viewer"
    METRICS_CHART = "metrics_chart"
    PROCESS_TABLE = "process_table"
    JSON_TREE = "json_tree"
    CODE_BLOCK = "code_block"
    DIFF_VIEW = "diff_view"
    TABLE = "table"
    TEXT = "text"
    CHART = "chart"


class ServerMessage(BaseModel):
    type: ServerMessageType
    timestamp: datetime = Field(default_factory=lambda: datetime.now(UTC))
    data: dict[str, Any] = Field(default_factory=dict)


class ActionRequest(BaseModel):
    """A proposed action that needs user approval."""

    id: str
    tool: str
    description: str
    command: list[str] | str
    risk_level: str  # READ_ONLY, LOW, MEDIUM, HIGH, CRITICAL
    reversible: bool = False
    requires_password: bool = False


class AlertMessage(BaseModel):
    """An alert notification sent to the client."""

    id: str
    severity: str  # CRITICAL, WARNING, INFO
    title: str
    summary: str
    source: str
    timestamp: datetime = Field(default_factory=lambda: datetime.now(UTC))
    investigation_thread_id: str | None = None
